internal/blockprocessor/validator.go

```diff
@@ -182,6 +190,40 @@ func (v *validator) validateBlock(block *types.Block) ([]*types.ValidationInfo,
 	}
 }
 
+func (v *validator) parallelSigValidation(dataTxEnvs []*types.DataTxEnvelope) ([]*types.ValidationInfo, [][]string, error) {
+	valInfoArray := make([]*types.ValidationInfo, len(dataTxEnvs))
+	usersWithValidSigPerTX := make([][]string, len(dataTxEnvs))
+	errorArray := make([]error, len(dataTxEnvs))
+
+	var wg sync.WaitGroup
+	wg.Add(len(dataTxEnvs))
+
+	for txNum, txEnv := range dataTxEnvs {
+		go func(txEnv *types.DataTxEnvelope, usersWithValidSig *[]string, valInfo **types.ValidationInfo, err *error) {
+			usersWithValidSignTx, vInfo, vErr := v.dataTxValidator.validateSignatures(txEnv)
+			if vErr != nil {
+				*err = vErr
+			} else {
+				*usersWithValidSig = usersWithValidSignTx
+				*valInfo = vInfo
+				if vInfo.Flag != types.Flag_VALID {
+					v.logger.Debugf("data transaction [%v] is invalid due to [%s]", txEnv.Payload, vInfo.ReasonIfInvalid)
+				}
+			}
+			wg.Done()
+		}(txEnv, &usersWithValidSigPerTX[txNum], &valInfoArray[txNum], &errorArray[txNum])
+	}
```

Comment:
The following can be used to improve code readability (it does not create any race condition too). 

```suggestion
		go func(txEnv *types.DataTxEnvelope, txNum int) {
			usersWithValidSignTx, vInfo, vErr := v.dataTxValidator.validateSignatures(txEnv)
			if vErr != nil {
				errorArray[txNum] = err
			} else {
				usersWithValidSigPerTx[txNum] = usersWithValidSignTx
				valInfoArray[txNum] = vInfo
				if vInfo.Flag != types.Flag_VALID {
					v.logger.Debugf("data transaction [%v] is invalid due to [%s]", txEnv.Payload, vInfo.ReasonIfInvalid)
				}
			}
			wg.Done()
		}(txEnv, txNum)
	}
```

Reply:
Nit: we can avoid the extra indentation. 

```
		go func(txEnv *types.DataTxEnvelope, txNum int) {
			defer wg.Done()
			
                        usersWithValidSignTx, vInfo, vErr := v.dataTxValidator.validateSignatures(txEnv)
			if vErr != nil {
				errorArray[txNum] = vErr
				return
			}

			usersWithValidSigPerTX[txNum] = usersWithValidSignTx
			valInfoArray[txNum] = vInfo
			if vInfo.Flag != types.Flag_VALID {
				v.logger.Debugf("data transaction [%v] is invalid due to [%s]", txEnv.Payload, vInfo.ReasonIfInvalid)
			}
		}(txEnv, txNum)

```

Reply:
Done